gitlabform/gitlabform/processors/group/group_members_processor.py

```diff
@@ -0,0 +1,121 @@
+
+                    if user in users_before_by_username:
+
+                        access_level_before = users_before_by_username[user][
+                            "access_level"
+                        ]
+                        expires_at_before = users_before_by_username[user]["expires_at"]
+
+                        if (
+                            access_level_before == access_level_to_set
+                            and expires_at_before == expires_at_to_set
+                        ):
+                            logging.debug(
+                                "Nothing to change for user '%s' - same config now as to set.",
+                                user,
+                            )
+                        else:
+                            logging.debug(
+                                "Re-adding user '%s' to change their access level or expires at.",
+                                user,
+                            )
+                            # we will remove the user first and then re-add they,
+                            # to ensure that the user has the expected access level
+                            self.gitlab.remove_member_from_group(group, user)
+                            self.gitlab.add_member_to_group(
+                                group, user, access_level_to_set, expires_at_to_set
+                            )
+
+                    else:
+                        logging.debug(
+                            "Adding user '%s' who previously was not a member.", user
+                        )
+                        self.gitlab.add_member_to_group(
+                            group, user, access_level_to_set, expires_at_to_set
+                        )
+
+            if configuration.get("enforce_group_members"):
+                # remove users not configured explicitly
+                # note: only direct members are removed - inherited are left
+                users_not_configured = set(
+                    [user["username"] for user in users_before]
+                ) - set(users_to_set_by_username.keys())
+                for user in users_not_configured:
+                    logging.debug(
+                        "Removing user '%s' who is not configured to be a member.", user
+                    )
+                    self.gitlab.remove_member_from_group(group, user)
+            else:
+                logging.debug("Not enforcing group members.")
+
+            logging.debug(
+                "Group members AFTER: %s", self.gitlab.get_group_members(group)
+            )
+
+        else:
+
+            logging.fatal(
+                "You cannot configure a group to have no members. GitLab requires a group "
+                " to contain at least 1 member who is an Owner (access_level = 50)."
+            )
+            sys.exit(4)
+
+    def _log_changes(self, group: str, group_members):
+        logging.info("Diffing for group_members section is not supported yet")
```

Comment:
Nit:

Can't we somehow put "Diffing for \<this\> section is not supported yet" in the `AbstractProcessor` class?

What we currently have feels a bit copy-pasty to me.

Reply:
You are right, though ultimately we do not want this to be printed anywhere. Moved to the `AbstractProcessor` and removed from classes which do not implement diffing (yet)